### PR TITLE
[ADD][product_properties_by_category]Settings up for product properties

### PR DESCRIPTION
**Pull request by hugho-ad**
[TASK#3652](https://www.vauxoo.com/#id=3652&view_type=form&model=project.task&menu_id=602&action=778)
[HU#888](https://www.vauxoo.com/#id=888&view_type=form&model=user.story) sprint 14

Add field selection to product.category to setting up the call for bids y product.tamplate

### CODE REVIEW

File: product_properties_by_category/README.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Product Properties by Category
+============================
+
+New functionality to configurate the propierties of Product Category
+
+A new section called "Product Properties" in the Product Category
+where can configure “Call for Bids” of the product.
```
Comment (zaoral) [inline product_properties_by_category/README.rst]: @hugho-ad please change the description content.
1. complete the line of the module title in line 2 to the same number of characters of the line 1. This need to match to be a valid rst syntax.
2. Change the description of the module with this:

```
This module add a new section called "Product Property" in the Product Category
form view to manage a group of common default values to be set at the Products
that belongs to the Category.

When you change the Internal Category of the Product the default values defined in
the Category for the Products will be set. This configuration is set but is not mandatory,
you can manually change the Products fields as you need.

For the current version, only works with the "Call For Bids" Product field.
When you update your Products Internal Category the "Call For Bids" field
in the Product will be updated with the default value given to the field
"Call For Bids" defined in the Product Category. If the Product Category
have not defined a default value for "Call For Bids" then will look into the
parent Product Category an so on until find a default value. If not default
value is set on any of the parent product category then the "Call For Bids"
field in Product will be set to False.
```

File: product_properties_by_category/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+###############################################################################
+
+{
+    "name": "Product Properties by Category",
+    "summary": "Set Properties by Category",
```
Comment (zaoral) [inline product_properties_by_category/__openerp__.py]: @hugho-ad change to 

```
-   "summary": "Set Properties by Category",
+   "summary": "Set Product Default Properties from the Product Category",
```

File: product_properties_by_category/i18n/es.po [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+"Language-Team: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: \n"
+"Plural-Forms: \n"
+
+#. module: product_properties_by_category
+#: selection:product.category,purchase_requisition:0
+msgid "Not Set"
+msgstr "No Configurado"
```
Comment (zaoral) [inline product_properties_by_category/i18n/es.po]: @hugho-ad "Sin Default". Apply the same to the es_PA translation

File: product_properties_by_category/model/pruduct_category.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    _inherit = 'product.template'
+
+    @api.onchange('categ_id')
+    def _get_prod_cate_call_of_bids(self):
+        """
+        Reviews the state of the field "Call of Bids" in the product.category
+        and update the state of the field "Call of Bids"
+        in the product.template (Boolean field with the same name).
+        If the field purchase_requisition product.category has "Not Set" option
+        then review the product category father
```
Comment (zaoral) [inline product_properties_by_category/model/pruduct_category.py]: @hugho-ad change "product category father" for "product category parent"

File: product_properties_by_category/model/pruduct_category.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        and so on until you find a category where you have defined
+         (other than 'Not Set') value.
+        """
+        self.purchase_requisition = self.check_parents_prch_rqs(self.categ_id)
+
+    @api.one
+    def check_parents_prch_rqs(self, categ_id):
```
Comment (zaoral) [inline product_properties_by_category/model/pruduct_category.py]: @hugho-ad this method is correct. but is a recursion so will bring computational cost. There is a different way yo fullfil this issue in a linear cost. I will search for an example and I will reference to you to make the change.

File: product_properties_by_category/model/pruduct_category.py (hunk 4 above)
Comment (nhomar) [inline product_properties_by_category/model/pruduct_category.py]: In fact is: parent product catgory

**Comment by zaoral** [comment]
@hugho-ad please change your file name model/pruduct_category.py. to model/product_category.py

**Comment by zaoral** [comment]
@hugho-ad please Hugo add a unit test to this module. Please use this schema:

```
1. Setup: Create o search a product that have a hierarchy product category.
    Product < Category A < Category B < Category C.
2. Test 01: Check that the product receive the default value defined in the product.category
2.1. Set "Call for Bids" False to the product.
2.2. Add a default value "Set True" for "Call for Bids" of product category first in line (Category A).
2.3. Simulate the onchange and check that the field in the product change for False to True.
3. Test 02: No restriction over the inherit values. The user can manually change a value inherit form the product.category as is needed.
3.1. Taking the same product in Test 01 change the "Call for Bids" in product to False.
3.2. Check the new value of product "call for bids" with the product.category A "call for bids" to check that they are different.
4. Test 03: Check pull the default value form the parent category one level above.
4.1. Remove the default value in Category A and add a default value "Set True" in Category B.
4.2. Simulate the onchane and check that de product "Call for Bids" change to default describe in Category B.
5. Test 04: What happen when is not default value defined in the product category or in either of its parents.
5.1. Change the product "Call for Bids" to True.
5.2. Remove all default in Category A/B/C. (Set then to "No set").
5.3. Simulate the onchange and check that the Call for Bids change form True to False.
```

**Comment by zaoral** [comment]
@hugho-ad please add a string and help to the field ProductCategory.purchase_requistion, use the ones defined in the [product](https://github.com/vauxoo/odoo/blob/8.0-vauxoo/addons/purchase_requisition/purchase_requisition.py#L387). For this maybe you will need to update the translation files too.

**Comment by zaoral** [comment]
@hugho-ad I create a pull request when you can find a diff of the changes I want you to test and apply to this PR vauxoo-dev/addons-vauxoo/pull/7, After finish the unit test please try to make merge into your branch.

**Comment by hugho-ad** [comment]
@zaoral ,I'm corrected, the function get_purchase_requisition_default by adding a break sentence
the tests was modified in order to your request.
Categoy A like the child category, category b like the parent and category C like parent of the parent

Regards

**Comment by zaoral** [comment]
+1

**Comment by zaoral** [comment]
@nhomar This PR is ready to review.

NOTE: The testing tools results: Runbot Green and we have a "False Red" for travis. This False Red is becuase a pylint error that currently is been in progress to be resolve for the vauxoo team.

**Pull request merged**